draft-ietf-quic-http.md

```diff
@@ -549,20 +549,38 @@ SETTINGS frame MUST be processed in the order they appear, with no other frame
 processing between values. Unsupported parameters MUST be ignored. 
 
 Once all values have been processed, if the REQUEST_ACK flag was set, the 
-recipient MUST immediately emit a SETTINGS_ACK frame listing the identifiers 
-whose values were understood and applied. (If none of the values were 
-understood, the SETTINGS_ACK frame will be empty, but MUST still be sent.) Upon 
-receiving an SETTINGS_ACK frame, the sender of the altered parameters can rely 
-on the setting having been applied. 
-
-TODO:
-: The above text was written for HTTP/2 -- QUIC has cross-stream timing issues 
-here that need to be solved.
-
+recipient MUST immediately emit the following:
+
+ - On the connection control stream, a SETTINGS_ACK frame listing the 
+   identifiers whose values were understood and applied. (If none of the values 
+   were understood, the SETTINGS_ACK frame will be empty, but MUST still be
+   sent.) 
+
+ - On each request control stream which is not in the "half-closed (local)" or
+   "closed" state, an empty SETTINGS_ACK frame.
+
+The SETTINGS_ACK frame on the connection control stream contains the highest
+stream number which was open at the time the SETTINGS frame was received.  All
+streams with higher numbers can safely be assumed to have the new settings in
+effect when they open.
```

Comment:
Let's say that the opening of a stream is delayed (loss looks like a delay for this stuff).  Stream N gets through, but stream N-2 takes another RTT or so to open.  What is the receiver of the SETTINGS frame expected to do in this case?  Is it expected to play out the SETTINGS_ACK when the stream finally makes it through?

The problem here is that stream N-2 might have been skipped by the other side.  How does a receiver know what to do with that uncertainty?  It can't hold onto the SETTINGS_ACK indefinitely.

I think that we need to mandate no stream skipping in this case.

The next problem is in correlating the SETTINGS_ACK with the SETTINGS.  I think that your scheme almost works, but we probably want to include some text on how this correlation is expected to work:
 - for every SETTINGS that requests acknowledgment, there is a SETTINGS_ACK
 - for every SETTINGS_ACK, every open stream less or equal to the included stream number will contain a SETTINGS_ACK

Where this doesn't work is when the receiver of a SETTINGS frame opens a stream.  Say a server sends SETTINGS at and then a client opens stream 7.  The server cannot know whether stream 7 was open at the time that the client received the SETTINGS frame.  Even if the opening frame on stream 7 is in the same packet as the client's SETTINGS_ACK, there is no good way to know what order the client actually performed these two actions.  Thus, if stream 7 is still open when a second SETTINGS is sent, the SETTINGS_ACK that the server receives on stream 7 could correspond to either settings.

The only fix I'm aware of is to set start AND end markers on the control stream SETTINGS_ACK.